File: 165 Compare Version Numbers/165.compare-version-numbers.py
#
# @lc app=leetcode id=165 lang=python3
#
# [165] Compare Version Numbers
#

# @lc code=start
class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        # 1. Split both version strings into lists of integers by the '.' delimiter
        v1 = list(map(int, version1.split('.')))
        v2 = list(map(int, version2.split('.')))

        # 2. Determine the maximum length of the two version lists
        max_len = max(len(v1), len(v2))

        # 3. Iterate through the revisions of both versions
        for i in range(max_len):
            # 4. Get the current revision number for each version or use 0 if index exceeds length
            r1 = v1[i] if i < len(v1) else 0
            r2 = v2[i] if i < len(v2) else 0

            # 5. Compare the current revisions
            if r1 < r2:
                return -1  # version1 < version2
            elif r1 > r2:
                return 1  # version1 > version2

        # 6. If all revisions are equal, return 0 (versions are equal)
        return 0
# @lc code=end

File: 165 Compare Version Numbers/165.compare-version-numbers.cpp
/*
 * @lc app=leetcode id=165 lang=cpp
 *
 * [165] Compare Version Numbers
 */

// @lc code=start
#include <vector>
#include <string>
#include <sstream>
#include <iostream>

class Solution {
public:
    int compareVersion(string version1, string version2) {
        // 1. Split both version strings into vectors of integers by the '.' delimiter
        vector<int> v1 = split(version1);
        vector<int> v2 = split(version2);

        // 2. Determine the maximum length of the two version vectors
        int max_len = max(v1.size(), v2.size());

        // 3. Iterate through the revisions of both versions
        for (int i = 0; i < max_len; i++) {
            // 4. Get the current revision number for each version or use 0 if index exceeds size
            int r1 = i < v1.size() ? v1[i] : 0;
            int r2 = i < v2.size() ? v2[i] : 0;

            // 5. Compare the current revisions
            if (r1 < r2) {
                return -1;  // version1 < version2
            } else if (r1 > r2) {
                return 1;  // version1 > version2
            }
        }

        // 6. If all revisions are equal, return 0 (versions are equal)
        return 0;
    }

private:
    // Utility function to split a version string by '.' and return vector of integers
    vector<int> split(const string& version) {
        vector<int> result;
        stringstream ss(version);
        string token;
        
        // Split by '.' and convert each part to an integer
        while (getline(ss, token, '.')) {
            result.push_back(stoi(token));
        }

        return result;
    }
};
// @lc code=end

